lib/routes/totp.js

```diff
@@ -0,0 +1,233 @@
+          .then(emitMetrics)
+          .then(createResponse)
+          .then(() => reply(response), reply)
+
+        const secret = otplib.authenticator.generateSecret()
+
+        function createTotpToken() {
+          if (sessionToken.tokenVerificationId) {
+            throw errors.unverifiedSession()
+          }
+
+          return db.createTotpToken(uid, secret, 0)
+        }
+
+        function createResponse() {
+          const otpauth = otplib.authenticator.keyuri(sessionToken.email, config.serviceName, secret)
+
+          return qrcode.toDataURL(otpauth, qrCodeOptions)
+            .then((qrCodeUrl) => {
+              response = {
+                qrCodeUrl,
+                secret
+              }
+            })
+        }
+
+        function emitMetrics() {
+          log.info({
+            op: 'totpToken.created',
+            uid: uid
+          })
+          return request.emitMetricsEvent('totpToken.created', {uid: uid})
+        }
+      }
+    },
+    {
+      method: 'POST',
+      path: '/totp/destroy',
+      config: {
+        auth: {
+          strategy: 'sessionToken'
+        },
+        response: {}
+      },
+      handler(request, reply) {
+        log.begin('totp.destroy', request)
+
+        const sessionToken = request.auth.credentials
+        const uid = sessionToken.uid
+
+        customs.check(request, 'totpDestroy')
+          .then(deleteTotpToken)
+          .then(() => reply({}), reply)
+
+
+        function deleteTotpToken() {
+          if (sessionToken.tokenVerificationId) {
+            throw errors.unverifiedSession()
+          }
+
+          return db.deleteTotpToken(uid)
+        }
+      }
+    },
+    {
+      method: 'GET',
+      path: '/totp/exists',
+      config: {
+        auth: {
+          strategy: 'sessionToken'
+        },
+        response: {
+          schema: isA.object({
+            exists: isA.boolean()
+          })
+        }
+      },
+      handler(request, reply) {
+        log.begin('totp.exists', request)
+
+        const sessionToken = request.auth.credentials
+        let exists = false
+
+        return getTotpToken()
+          .then(() => reply({exists}), reply)
+
+        function getTotpToken() {
+          if (sessionToken.tokenVerificationId) {
+            throw errors.unverifiedSession()
+          }
+
+          return db.totpToken(sessionToken.uid)
+            .then((token) => {
+              if (token) {
+                exists = true
+              }
+            }, (err) => {
+              if (err.errno === errors.ERRNO.TOTP_TOKEN_NOT_FOUND) {
+                exists = false
+                return
+              }
+              throw err
+            })
+        }
+      }
+    },
+    {
+      method: 'POST',
+      path: '/session/verify/totp',
+      config: {
+        auth: {
+          strategy: 'sessionToken'
+        },
+        validate: {
+          payload: {
+            code: isA.string().max(32).required(),
```

Comment:
@vbudhram can we validate these as digits/hex/alphanumeric?

Reply:
Looks like this has been fixed recently and only accepts digits.